### ISSUE TITLE
Nullish coalescing operator corrupts for-of loop when targeting ES5

### ISSUE THREAD

**Issue by dmester** [microsoft/TypeScript#39907]
**TypeScript Version:** 3.7.5 to 4.0.0-dev.20200803

**Search Terms:**
ES5
Optional chaining operator
Nullish coalescing operator
"Cannot read property 'next' of undefined"
"Cannot read property 'length' of undefined"

**Code**
```ts
const outerItems: any = [
    { innerItems: [ {} ] }
];

for (const outerItem of outerItems) {
    () => {
        outerItem;
    };
    for (const innerItem of outerItem.innerItems) {
        innerItem.member ?? "";
    }
}

console.log("Succeeded");
```

Is compiled to (target ES5):

```js
"use strict";
var _a;
var outerItems = [
    { innerItems: [{}] }
];
var _loop_1 = function (outerItem) {
    (function () {
        outerItem;
    });
    for (var _i = 0, _a = outerItem.innerItems; _i < _a.length; _i++) {
        var innerItem = _a[_i];
        (_a = innerItem.member) !== null && _a !== void 0 ? _a : "";
    }
};
for (var _i = 0, outerItems_1 = outerItems; _i < outerItems_1.length; _i++) {
    var outerItem = outerItems_1[_i];
    _loop_1(outerItem);
}
console.log("Succeeded");
```

**Expected behavior:**
Output: 
```
Succeeded
```

**Actual behavior:**
Error is thrown when compiled output is executed:
```
Cannot read property 'length' of undefined 
```

Temporary variable `_a` holding the iterated array is overwritten by the nullish coalescing operator.

The state is also corrupted if:
* `??` is replaced with `?.`  [Playground](https://www.typescriptlang.org/play?noImplicitAny=false&target=1&ts=4.0.0-dev.20200803#code/MYewdgzgLgBCCuUCmAnAksgthAXDAhmAJ4wC8MA2gLABQM9MA3jAJZhioZLZ4VMC+MALox+tIQG5atAGYgUMABShIsBMnRY4MuIk5YIASia0GS46QB8JumYbr93Kbfr9nZuQuXhordo8xtXQ0uTAA6Ng5NbiMbOwZIgLDMbgAjVAB+MJAoAAtUAFk01HcGMRpy2hUIEAAbJDDakABzRQAiAGV4YGAkJAATAbbDCSA)
* `downlevelIteration` is enabled, where it will instead overwrite the variable holding the iterator. [Playground](https://www.typescriptlang.org/play?noImplicitAny=false&downlevelIteration=true&target=1&ts=4.0.0-dev.20200803#code/MYewdgzgLgBCCuUCmAnAksgthAXDAhmAJ4wC8MA2gLABQM9MA3jAJZhioZLZ4VMC+MALox+tIQG5atAGYgUMABShIsBMnRY4MuIk5YIASia0GS46QB8JumYbr93Kbfr9nZuQuXhordo8xtXQ0uTAA6Ng5NbiMbOwZIgLDMbgAjVBgAfkyYACJc9wYxGmLaFQgQABskMMqQAHNFXIBleGBgJCQAE27cwwkgA)

The emulated block scope for ES5 caused by the inner function seems to be the key for triggering this behavior.

**Playground Link:** 
[Playground](https://www.typescriptlang.org/play?noImplicitAny=false&target=1&ts=4.0.0-dev.20200803#code/MYewdgzgLgBCCuUCmAnAksgthAXDAhmAJ4wC8MA2gLABQM9MA3jAJZhioZLZ4VMC+MALox+tIQG5atAGYgUMABShIsBMnRY4MuIk5YIASia0GS46QB8JumYbr93Kbfr9nZuQuXhordo8xtXQ0uTAA6Ng5NbiMbOwZIgLDMbgAjVBgAfkyYACJc9wYxGmLaFQgQABskMMqQAHNFXIBleGBgJCQAE27cwwkgA)

**Related Issues:** 
https://github.com/microsoft/TypeScript/issues/37686

**Comment by rbuckton** [comment]
```ts repro
// @showEmit
// @target: es5
const outerItems: any = [
    { innerItems: [ {} ] }
];

for (const outerItem of outerItems) {
    () => {
        outerItem;
    };
    for (const innerItem of outerItem.innerItems) {
        innerItem.member ?? "";
    }
}

console.log("Succeeded");
```

[Workbench Repro](https://www.typescriptlang.org/dev/bug-workbench/?target=1&ts=4.2.0-dev.20201211#code/PTAEAEGcAsHsHcCiBbAlgFwLACgQXQIYBOA5gKboBcoZkArDgMawB2k6osArumUQJK9kkagRYBPUAF5QAbRyhFoAN6hULFn0Flh1WSoC+oALqgDOYwG4cOAGawioABTM2HbrwFDOtzjy1CkACUKgpKTiFSAHyh2Erxfp7ayNZxSgap8faOLqzsahoBOj6JRcgAdOqaXjrBsQlKVWXlyDoARnygAPxdoABEfZnpOObYTHmwADZk5ZOwJE59AMpcjIxkZAAmW31BqUA)

**Comment by rbuckton** [comment]
Fixed by #41791
